### PR TITLE
[cling] Rewrite implementation of `InputValidator::validate()`; fixes ROOT-9202

### PR DESCRIPTION
**Pull request by jalopezg-r00t**
This pull request replaces the implementation of `InputValidator::validate()` by simpler, more maintainable code that also fixes JIRA issue [ROOT-9202](https://sft.its.cern.ch/jira/browse/ROOT-9202).

The previous implementation was unable to properly handle line continuation after ',' or '\\'.  Specifically, it skipped over non-punctuation tokens, entering continuation mode even if ',' or '\' were not the last tokens in the input, e.g.
```
int a, b
```
or
```
int a \      b
```
caused cling to request more input, where it shouldn't.

## Changes or fixes:
- MetaLexer:
  - Return `/*` and `*/` as independent tokens.
  - Added `ReadToEndOfLine()` function (consume input until '\r', '\n', or EOF).
  - Added `MetaLexer::RAII` that saves the current lexing position and restores it on destruction.
  - Remove unused `LexPunctuatorAndAdvance()`.
- Rewrite of `InputValidator::validate()`.

## Checklist:
- [X] tested changes locally

Fixes [ROOT-9202](https://sft.its.cern.ch/jira/browse/ROOT-9202).

### CODE REVIEW

**Comment by jalopezg-r00t** [comment]
@Axel-Naumann, @vgvassilev, as I will be off in August, again feel free to merge this if, after the review, it can be merged right away.

File: interpreter/cling/lib/MetaProcessor/InputValidator.cpp
Comment (hahnjo) [inline interpreter/cling/lib/MetaProcessor/InputValidator.cpp:90]: Isn't this just `PPtk.size() == 5 || PPtk[5]=='/' || isspace(PPtk[5])`, ie exact match or `/` or space? Why could there be a slash?

File: interpreter/cling/include/cling/MetaProcessor/MetaLexer.h
Comment (jalopezg-r00t) [reply interpreter/cling/include/cling/MetaProcessor/MetaLexer.h:83]: Added a comment there.  Thanks!

File: interpreter/cling/lib/MetaProcessor/InputValidator.cpp
Comment (jalopezg-r00t) [reply interpreter/cling/lib/MetaProcessor/InputValidator.cpp]: That's right!  It was inherited from the old code and I didn't change.  ;-)

File: interpreter/cling/lib/MetaProcessor/InputValidator.cpp
Comment (jalopezg-r00t) [reply interpreter/cling/lib/MetaProcessor/InputValidator.cpp]: Thanks for spotting, @hahnjo!  Again, from the old code...  ;-)

File: interpreter/cling/include/cling/MetaProcessor/MetaLexer.h
Comment (vgvassilev) [inline interpreter/cling/include/cling/MetaProcessor/MetaLexer.h:83]: The indentation seems awkward. Can you split it in multiple lines?

File: interpreter/cling/lib/MetaProcessor/InputValidator.cpp
Comment (vgvassilev) [inline interpreter/cling/lib/MetaProcessor/InputValidator.cpp:57]: Do we want to add `LLVM_FALLTHROUGH` to silence potential warnings?

**Review by vgvassilev — COMMENTED** [review]
Thanks for working on this!

File: interpreter/cling/include/cling/MetaProcessor/MetaLexer.h
Comment (jalopezg-r00t) [reply interpreter/cling/include/cling/MetaProcessor/MetaLexer.h:83]: Fixed.

File: interpreter/cling/lib/MetaProcessor/InputValidator.cpp
Comment (jalopezg-r00t) [reply interpreter/cling/lib/MetaProcessor/InputValidator.cpp:57]: IMHO, it should not be needed, as there are no statements interleaved with the case labels.
Thanks for the review, @vgvassilev!

**Review by Axel-Naumann — APPROVED** [review]
Thanks!

**Comment by jalopezg-r00t** [comment]
@phsft-bot build

**Comment by jalopezg-r00t** [comment]
@phsft-bot build just on mac11.0/cxx17

**Pull request merged**
